armi/runLog.py

```diff
@@ -226,265 +195,346 @@ def setVerbosity(self, levelInput):
+        if "mpiRank" in kwargs:
+            mpiRank = int(kwargs["mpiRank"])
+            del kwargs["mpiRank"]
         else:
-            # for shame. Might make sense to move the styles stuff back into the
-            # Framework
-            from tparmi.gui.styles import dialogues
-
-            dlg = dialogues.ScrolledMessageDialog(None, msg, "Prompt")
-        response = dlg.ShowModal()
-        dlg.Destroy()
-        if response == wx.ID_CANCEL:
-            raise exceptions.RunLogPromptCancel("Manual cancellation of GUI prompt")
-        return response in [wx.ID_OK, wx.ID_YES]
-
-    elif context.CURRENT_MODE == Mode.INTERACTIVE:
-        response = ""
-        responses = [
-            opt for opt in options if opt in ["YES_NO", "YES", "NO", "CANCEL", "OK"]
-        ]
-
-        if "YES_NO" in responses:
-            index = responses.index("YES_NO")
-            responses[index] = "NO"
-            responses.insert(index, "YES")
-
-        if not any(responses):
-            raise RuntimeError("No suitable responses in {}".format(responses))
-
-        # highly requested shorthand responses
-        if "YES" in responses:
-            responses.append("Y")
-        if "NO" in responses:
-            responses.append("N")
-
-        while response not in responses:
-            LOG.standardLogMsg("prompt", statement)
-            LOG.standardLogMsg(
-                "prompt", "{} ({}): ".format(question, ", ".join(responses))
-            )
-            response = sys.stdin.readline().strip().upper()
+            mpiRank = context.MPI_RANK
 
-        if response == "CANCEL":
-            raise exceptions.RunLogPromptCancel(
-                "Manual cancellation of interactive prompt"
-            )
+        super(RunLogger, self).__init__(*args, **kwargs)
+        self.allowStopDuplicates()
 
-        return response in ["YES", "Y", "OK"]
+        if mpiRank == 0:
+            h = logging.StreamHandler()
+            h.setLevel(logging.INFO)
+            self.setLevel(logging.INFO)
+        else:
+            filePath = os.path.join("logs", RunLog.STDOUT_NAME.format(args[0], mpiRank))
+            h = logging.FileHandler(filePath)
+            h.setLevel(logging.WARNING)
+            self.setLevel(logging.WARNING)
 
-    else:
-        raise exceptions.RunLogPromptUnresolvable(
-            "Incorrect CURRENT_MODE for prompting user: {}".format(context.CURRENT_MODE)
-        )
+        form = logging.Formatter(RunLogger.FMT)
+        h.setFormatter(form)
+        self.addHandler(h)
 
+    def _log(self, *args, **kwargs):
+        """wrapper around the standard library Logger._log() method
 
-def warningReport():
-    LOG.warningReport()
+        The primary goal here is to allow us to support the deduplication of warnings.
+        NOTE: All of the *args and **kwargs logic here are mandatory, as the standard library implementation of this
+        method has been changing the number of kwargs between Python v3.4 and v3.9.
```

Comment:
Do we mostly just need this as a trap door out of the `single` filter? A perhaps simpler alternative could be the `extra` dict , which can be passed to the various `logging` functions to staple things to the log record's `__dict__`, which can then be used in the `Filter`. So when calling the module-global `runLog.log()`, it can do something like:
```python
    def log(self, msgType, msg, single=False, label=None):
        ...
        # Do the actual logging
        self.logger.log(msgLevel, msg, extra={"single": single})
```

Then on the Filter side, we would peek at the record with `getattr()` to access `single` and handle accordingly.

Thoughts?

Reply:
Yeah, I like your idea of using `extra`.  It's much cleaner.

Nice, man.